src/builder/QuarkBuilder.sol

```diff
@@ -209,46 +232,54 @@ contract QuarkBuilder {
-        uint256 amount,
+        address sender,
         address recipient,
-        Payment calldata payment,
+        uint256 amount,
+        string memory assetSymbol,
+        Payment memory payment,
         ChainAccounts[] calldata chainAccountsList
     ) external pure returns (BuilderResult memory) {
-        AssetPositionsWithChainId[] transferAssetPositions = filterAssetPositions(assetSymbol, chainAccountsList);
-        AssetPositionsWithChainId[] paymentAssetPositions;
-        if (payment.isToken) {
-            paymentAssetPositions = filterAssetPositions(payment.currency, chainAccountsList);
-        }
+        ChainAccounts[] memory transferChainAccounts = filterChainAccounts(assetSymbol, chainAccountsList);
 
         // INSUFFICIENT_FUNDS
         // There are not enough aggregate funds on all chains to fulfill the transfer.
-        uint256 aggregateTransferAssetBalance;
-        for (uint i = 0; i < transferAssetPositions.length; ++i) {
-            for (uint j = 0; j < transferAssetPositions[i].accountBalances.length; ++j) {
-                aggregateTransferAssetBalance += transferAssetPositions[i].accountBalances[j].balance;
+        {
+            uint256 aggregateTransferAssetBalance;
+            for (uint256 i = 0; i < transferChainAccounts.length; ++i) {
+                aggregateTransferAssetBalance +=
+                    sumBalances(findAssetPositions(assetSymbol, transferChainAccounts[i].assetPositionsList));
+            }
+            if (aggregateTransferAssetBalance < amount) {
+                revert InsufficientFunds();
             }
-        }
-        if (aggregateTransferAssetBalance < amount) {
-            revert InsufficientFunds();
         }
 
         // TODO: Pay with bridged payment.currency?
         // MAX_COST_TOO_HIGH
         // There is at least one chain that does not have sufficient payment assets to cover the maxCost for that chain.
         // Note: This check assumes we will not be bridging payment tokens for the user
-        if (payment.isToken) {
-            for (uint i = 0; i < payment.maxCosts.length; ++i) {
-                PaymentMaxCost memory maxCost = payment.maxCosts[i];
-                AssetPositionsWithChainId memory paymentAssetPosition = getAssetPositionsForSymbolAndChain(payment.currency, maxCost.chainId, chainAccountsList);
-                uint256 paymentAssetBalanceOnChain = sumBalances(paymentAssetPosition);
-                uint256 paymentAssetNeeded = maxCost.amount;
-                // If the payment token is the transfer token and this is the target chain, we need to account for the transfer amount when checking token balances
-                if (Strings.stringEqIgnoreCase(payment.currency, assetSymbol) && chainId == maxCost.chainId) {
-                    paymentAssetNeeded += amount;
+        {
+            if (payment.isToken) {
+                ChainAccounts[] memory paymentChainAccounts;
+                if (payment.isToken) {
```

Comment:
This check can be removed. We already check 2 lines up.

Reply:
good catch, we also don't need to have `filter` here or anywhere anymore.